### PR TITLE
added run_async_app, fixes #2375

### LINKED ISSUE #2375: add a way to run an app asynchron
## Long story short

It would be nice to have an async / coroutine that will take an app and run it. This is very usefull if you have an long running that wants to expose a rest interface.

## Steps to reproduce

go to the web.py file and look for run function that runs an app asynchron

### PR DESCRIPTION
**Pull request by hwmrocker**
## Checklist

- [ ] I think the code is well written
- [ ] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [ ] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names. 
- [ ] Add a new news fragment into the `changes` folder
  * name it `<issue_id>.<type>` for example (588.bug)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

### CODE REVIEW

File: aiohttp/web.py [hunk 1]
```diff
@@ -425,6 +425,50 @@ def _make_server_creators(handler, *, loop, ssl_context,
             uris.append(str(base_url.with_host(host).with_port(port)))
     return server_creations, uris
 
+@asyncio.coroutine
+def run_async_app(app, *, host=None, port=None, path=None, sock=None,
+            shutdown_timeout=60.0, ssl_context=None,
+            print=print, backlog=128, access_log_format=None,
+access_log=access_logger, handle_signals=True, loop=None):
+    yield from app.startup()
```
Comment (oleksandr-kuzmenko) [inline aiohttp/web.py]: now python3.4 is dropped and you should to use async / await syntax

**Comment by asvetlov** [comment]
How would you use `run_async_app()`?
Please provide a small snippet. No need PR modification, a comment would be enough.

**Comment by hwmrocker** [comment]
here is an example how i would run it:

```loop = asyncio.get_event_loop()
tasks = [
    asyncio.ensure_future(web.run_async_app(app, loop=loop)),
    asyncio.ensure_future(other_async_function()),
]
loop.run_until_complete(asyncio.gather(*tasks))
```

I will write some tests in the next days

File: aiohttp/web.py [hunk 2]
```diff
@@ -426,6 +426,51 @@ def _make_server_creators(handler, *, loop, ssl_context,
     return server_creations, uris
 
 
+async def run_async_app(app, *, host=None, port=None, path=None, sock=None,
```
Comment (jettify) [inline aiohttp/web.py:429]: This code very similar what we have in `run_app`, should we somehow reuse this logic in `run_app`, example is not correct but still:
```python
def run_app(...):
    loop.run_until_complet(run_async_app(...))
```

**Comment by samuelcolvin** [comment]
@hwmrocker can you give some more information about what "other_async_function" is and what it's doing? This is a code example, not really a use case.

My guess is that you would have a better time starting `other_async_function` as a task via `on_startup`.

**Comment by jettify** [comment]
I use similar approach for testing, run server in background while hitting it with client.

https://github.com/aio-libs/aiozipkin/blob/a1a239d6f5a42fce35ecc9810c09eb4ac1d89780/examples/aiohttp_example.py#L45-L66

**Comment by jettify** [comment]
Other use case, I have 5 services I want to run all of them in one event loop locally to simplify testing, 
https://github.com/aio-libs/aiozipkin/blob/a1a239d6f5a42fce35ecc9810c09eb4ac1d89780/demos/microservices/runner.py#L1-L33

Otherwise I have to run each service in separate thread or copy paste `run_app` in order make startup/shutdown logic work.

**Comment by asvetlov** [comment]
Well, I understand the motivation pretty clean but asked about not only spawning a new task but it cancellation also.
https://github.com/aio-libs/aiohttp/pull/2376/files#diff-8e4ccf3aaf43100064462a4df10952e5R460 looks pretty incorrect: `KeyboardInterrupt` will never appear.
I pretty sure making a class for handling all gory details as proposed by #2121 is much better solution.

@hwmrocker and others: would you change your vision by providing a new PR?
I remember @cecton was interesting on the feature too.

The issue is in my priority list anyway but I would be very happy if somebody make it done.

**Comment by cecton** [comment]
@asvetlov the motivation is unclear to me so I asked on #2375 for details.

To me it seems that the web service is the long running task of your process and your process should end with it. Unless you have a scenario where you want to have your process continue running without the web service, I think you can do everything with aiojobs (or without) and startup/clean hooks.

**Pull request closed without merging**